Azure provisioning in the dashboard

davidfowl opened the pull request:
This change makes it possible (and cleaner IMO) to publish and log data about a resource to the dashboard and does that to show errors when creating resources and for showing provisioning progress when using the **Aspire.Hosting.Azure.Provisioning** package.  This change consolidates the previously made annotations for publishing and subscribing to both resource updates and logs general purpose services instead of annotations. 

TODO:
- [x] - Make all azure resources work like this (I only did cosmosdb so far)
- [x] - Add some more tests for logic in ResourceNotificationService.

Fixes #2396 
Fixes #2569
Contributes to https://github.com/dotnet/aspire/issues/1758

Here's the bicep example:

<img width="1200" alt="image" src="https://github.com/dotnet/aspire/assets/95136/73271a0d-d88c-487e-8009-bae74970bd60">


 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2552)

davidfowl committed:
WIP

Show updates in UI from config

Remove calls from resource

Test changes

Show connection strings

WIP

Tweaks to make it work
- Return a snapshot on watch

Remove lifecycle hook

Fixed playground app

Fix tests

Remove connection strings until we get masking

Cleanup

Smaller diff

Cleanup

Small cleanup

Send updates to dashboard when resources fail to start

Post updates using NotificationService from the ApplicationExecutor

davidfowl committed:
More cleanup

davidfowl committed:
Remove comment

davidfowl commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -59,7 +60,9 @@ internal sealed class ApplicationExecutor(ILogger<ApplicationExecutor> logger,
                                           IDashboardEndpointProvider dashboardEndpointProvider,
                                           IDashboardAvailability dashboardAvailability,
-                                          DistributedApplicationExecutionContext executionContext)
+                                          DistributedApplicationExecutionContext executionContext,
```
These changes look more complex than they are. The main changes are moving `Create{Resource}Async` out to a single method so that multiple resources can be started allocated in parallel. We also publish status updates for starting and failure (if it fails to start).

davidfowl committed:
Always show the resource in the dashboard
- Added logging when configuration is missing from azure provisioner

davidfowl committed:
More tests

davidfowl committed:
There's no connection string support yet

mitchdenny commented on playground/CustomResources/CustomResources.AppHost/TestResource.cs:
```diff
@@ -28,53 +27,44 @@ public static IResourceBuilder<TestResource> AddTestResource(this IDistributedAp
-                    while (await timer.WaitForNextTickAsync(_tokenSource.Token))
-                    {
-                        var randomState = states[Random.Shared.Next(0, states.Length)];
+                using var timer = new PeriodicTimer(TimeSpan.FromSeconds(seconds));
```
You should pass the cancellation token into this.

mitchdenny commented on src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:
```diff
@@ -64,7 +64,7 @@ protected override void OnInitialized()
             new KnownProperty(KnownProperties.Resource.State, Loc[Resources.Resources.ResourcesDetailsStateProperty]),
             new KnownProperty(KnownProperties.Resource.CreateTime, Loc[Resources.Resources.ResourcesDetailsStartTimeProperty]),
-            new KnownProperty(KnownProperties.Resource.ExitCode, Loc[Resources.Resources.ResourcesDetailsExitCodeProperty])
+            new KnownProperty(KnownProperties.Resource.ExitCode, Loc[Resources.Resources.ResourcesDetailsExitCodeProperty]),
```
nit: trailing comma.

davidfowl commented:
/azp run

davidfowl committed:
Small tweaks

JamesNK commented on playground/CustomResources/CustomResources.AppHost/TestResource.cs:
```diff
@@ -28,53 +27,44 @@ public static IResourceBuilder<TestResource> AddTestResource(this IDistributedAp
 }
 
-internal sealed class TestResourceLifecycleHook : IDistributedApplicationLifecycleHook, IAsyncDisposable
+internal sealed class TestResourceLifecycleHook(ResourceNotificationService notificationService, ResourceLoggerService loggerService) : IDistributedApplicationLifecycleHook, IAsyncDisposable
```
I see that `AddTestResource` calls `builder.Services.AddLifecycleHook<TestResourceLifecycleHook>()`. What happens each time that call is made? Will it register multiple? Should it call TryAdd instead?

FYI `AddAzureProvisioning` calls `AddLifecycleHook`. Should it call TryAdd instead?

davidfowl replied:
Should be try adds.

davidfowl replied:
Fixed

davidfowl committed:
Add support for all azure resources
- Ran the bicep playground as a stress test
- Fixed bugs with not waiting properly
- Fixed issues with shownig errors after failed provisions
- Delete connection string commented out code

davidfowl committed:
Success case

davidfowl committed:
Fixed who calls trysetresult

davidfowl committed:
Make bicep value resolution async
- Fixed sample

davidfowl committed:
Log getting bicep values
- Remove resouce id check

davidfowl committed:
Add dashboard resource filtering

davidfowl committed:
Delete the placeholder before we add the update
